MT part tech trading + scrapping fixes

Bertie690 opened the pull request:
TODO:
- [x] Move code to its own function because both scenarions (scrapping/battle) use nearly the exact same code (i quite literally copypasted the entire thing and hit F2 a couple times)
- [x] Add tests for MT part trading 
- [x] Add messages for it to frontend bc I forgot

Fixes #414, #205, #470

Bertie690 committed:
Updated tech decs & fixed typo

Bertie690 committed:
Update mineralpacket.go

made random number ordering consistent betwen funcs

Bertie690 committed:
Added MT part tech trading

needs testing and stuff

Bertie690 committed:
fixed typo in README.md

sirgwain commented on cs/message.go:
```diff
@@ -171,6 +171,8 @@ const (
 	PlayerMessagePlayerTechLevelGainedInvasion
 	PlayerMessagePlayerTechLevelGainedScrapFleet
 	PlayerMessagePlayerTechLevelGainedBattle
+	PlayerMessagePlayerMysteryTechGainedScrapFleet
+	PlayerMessagePlayerMysteryTechGainedBattle
```
unfortunately, these have to be added at the end because they are incremented numbers. If you add them in the middle, all existing games will have messageIds that are incorrect.

At some point I'll convert these message types to strings so I can sort them, but that would involve a relatively complicated database upgrade step (because they're ints in the database as well...)

Bertie690 replied:
Moved em to the end.

sirgwain commented on cs/player.go:
```diff
@@ -42,6 +42,7 @@ type Player struct {
 	Spec                      PlayerSpec           `json:"spec,omitempty"`
 	leftoverResources         int
 	techLevelGained           bool
+	mtPartGained              bool
```
please make it more generic for future mods. `acquirablePartGained`

Bertie690 replied:
What should the names of the messages be?
Should it be `TechPartGained`, `TechComponentGained`, `AcquirableTechGained` or something else entirely?

sirgwain replied:
`PlayerMessagePlayerTechPartGained`

Bertie690 replied:
Wil do find and replace when i have time on computer

sirgwain committed:
Merge branch 'develop' of github.com:sirgwain/craig-stars into fixes

sirgwain committed:
fixed merge issues

Bertie690 committed:
fixed things and reworked everything

oh my god i finally finished this my lord help me aaaaaaaa

Bertie690 committed:
Merge branch 'develop' into fixes

sirgwain committed:
fixed a couple nil pointer issues

Bertie690 committed:
Minor fixes

Bertie690 committed:
Merge remote-tracking branch 'upstream/develop' into fixes

Bertie690 committed:
replaced empty fallthroughs code-wide with comma separation for consistency

also added mystery trader minelayer support (me want big fat rainbow minelayer)

Bertie690 committed:
replaced empty fallthroughs with comma separation for consistency

worst comes to worst we use newlines for separation

Bertie690 committed:
Merge branch 'fixes' of https://github.com/Bertie690/craig-stars into fixes

Bertie690 committed:
Update planet.go

Bertie690 committed:
Added tests; fixed stuff

Bertie690 committed:
Added wrapper functions

Bertie690 committed:
Merge branch 'develop' into fixes

sirgwain committed:
added client side descriptions to TechGained messages

sirgwain committed:
regenerated rules json tags

sirgwain committed:
regenerated wasm

sirgwain committed:
added playerAquiredTech: true test case to ensure we don't aquire multiple parts in the same turn

Bertie690 committed:
fix
